src/main/kotlin/io/libp2p/core/security/noise/NoiseXXSecureChannel.kt

```diff
@@ -0,0 +1,207 @@
+import io.netty.channel.SimpleChannelInboundHandler
+import org.apache.logging.log4j.Level
+import org.apache.logging.log4j.LogManager
+import org.apache.logging.log4j.core.config.Configurator
+import spipe.pb.Spipe
+import java.util.concurrent.CompletableFuture
+
+class NoiseXXSecureChannel(val localKey: PrivKey, val localDHState: DHState, val remoteDHState: DHState, val role: Int) :
+    SecureChannel {
+    private val logger = LogManager.getLogger(NoiseXXSecureChannel::class.java.name + ":" + role)
+
+    private val handshakeHandlerName = "NoiseHandshake"
+
+    companion object {
+        const val protocolName = "Noise_XX_25519_ChaChaPoly_SHA256"
+        const val announce = "/noise/$protocolName/0.1.0"
+    }
+
+    override val announce = Companion.announce
+    override val matcher = ProtocolMatcher(Mode.PREFIX, name = "/noise/$protocolName/0.1.0")
+
+    init {
+        Configurator.setLevel(NoiseXXSecureChannel::class.java.name + ":" + role, Level.DEBUG)
+    }
+
+    fun initChannel(ch: P2PAbstractChannel): CompletableFuture<SecureChannel.Session> {
+        return initChannel(ch, "")
+    }
+
+    override fun initChannel(ch: P2PAbstractChannel, selectedProtocol: String): CompletableFuture<SecureChannel.Session> {
+        val ret = CompletableFuture<SecureChannel.Session>()
+        val resultHandler = object : ChannelInboundHandlerAdapter() {
+            override fun userEventTriggered(ctx: ChannelHandlerContext, evt: Any) {
+                when (evt) {
+                    is SecureChannelInitialized -> {
+                        ctx.channel().attr(SECURE_SESSION).set(evt.session)
+                        ret.complete(evt.session)
+                        ctx.pipeline().remove(handshakeHandlerName)
+                        ctx.pipeline().remove(this)
+                        logger.debug("Reporting secure channel initialized")
+                    }
+                    is SecureChannelFailed -> {
+                        ret.completeExceptionally(evt.exception)
+                        ctx.pipeline().remove(handshakeHandlerName)
+                        ctx.pipeline().remove(this)
+                        logger.debug("Reporting secure channel failed")
+                    }
+                }
+                ctx.fireUserEventTriggered(evt)
+            }
+        }
+        ch.nettyChannel.pipeline().addLast(handshakeHandlerName, NoiseIoHandshake())
+        ch.nettyChannel.pipeline().addLast(handshakeHandlerName + "ResultHandler", resultHandler)
+        return ret
+    }
+
+    inner class NoiseIoHandshake : SimpleChannelInboundHandler<ByteBuf>() {
+        private val handshakestate: HandshakeState = HandshakeState(protocolName, role)
+
+        init {
+            handshakestate.localKeyPair.copyFrom(localDHState)
+            handshakestate.start()
+            logger.debug("Starting handshake")
+        }
+
+        override fun channelRead0(ctx: ChannelHandlerContext?, msg: ByteBuf?) {
```

Comment:
When you extending `SimpleChannelInboundHandler` all the read handling should be made within this method and only `ByteBuf` instances are the messages (no `byte[]` are expected here)
Nor `super` call neither `channelRead` calls are expected here.

Reply:
done